File: LeetCode/java/2094_Finding_3-Digit_Even_Numbers.java
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

class Solution {
    void dfs(int cnt, int[] digits, boolean[] visit, List<Integer> list, Set<Integer> set) {
        if (cnt == 3) {
            if (list.get(0) == 0 || list.get(2) % 2 != 0) {
                return;
            }

            int num = 0;
            for (int n : list) {
                num = num * 10 + n;
            }

            set.add(num);
        } else {
            for (int i = 0; i < digits.length; i++) {
                if (visit[i] == true) {
                    continue;
                }
                visit[i] = true;
                list.add(digits[i]);
                dfs(cnt + 1, digits, visit, list, set);
                list.remove(list.size() - 1);
                visit[i] = false;
            }
        }
    }

    public int[] findEvenNumbers(int[] digits) {

        boolean[] visit = new boolean[digits.length];
        List<Integer> list = new ArrayList<>();
        Set<Integer> set = new TreeSet<>();

        dfs(0, digits, visit, list, set);

        int[] answer = new int[set.size()];

        int idx = 0;
        for (int n : set) {
            answer[idx++] = n;
        }

        return answer;
    }
}

File: LeetCode/cpp/2094_Finding_3-Digit_Even_Numbers.cpp
class Solution {
public:
	void dfs(int cnt, vector<int>& digits, vector<bool>& visit, vector<int>& v, set<int>& st) {
		if (cnt == 3) {
			if (v[0] == 0 || v[2] % 2 != 0) {
				return;
			}

			int num = 0;
			for (int n : v) {
				num = num * 10 + n;
			}

			st.insert(num);
		}
		else {
			for (int i = 0; i < digits.size(); i++) {
				if (visit[i] == true) {
					continue;
				}
				visit[i] = true;
				v.push_back(digits[i]);
				dfs(cnt + 1, digits, visit, v, st);
				v.pop_back();
				visit[i] = false;
			}
		}
	}

	vector<int> findEvenNumbers(vector<int>& digits) {
		vector<int> answer;

		vector<bool> visit(digits.size(), false);
		vector<int> v;
		set<int> st;

		dfs(0, digits, visit, v, st);

		for (int n : st) {
			answer.push_back(n);
		}

		return answer;
	}
};